extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -0,0 +1,706 @@
+     */
+    this.stylesheets_ = [];
+
+    /** @private {IntersectionObserver} */
+    this.intersectionObserver_ = null;
+  }
+
+  /** @override */
+  getPriority() {
+    // Priority used for scheduling preload and layout callback.  Because
+    // AMP creatives will be injected as part of the promise chain created
+    // within onLayoutMeasure, this is only relevant to non-AMP creatives
+    // therefore we want this to match the 3p priority.
+    return 2;
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  renderOutsideViewport() {
+    // Only relevant if non-AMP as AMP creative will be injected within
+    // buildCallback promise chain.
+    // If another ad is currently loading we only load ads that are currently
+    // in viewport.
+    const allowRender = allowRenderOutsideViewport(this.element, this.getWin());
+    if (allowRender !== true) {
+      return allowRender;
+    }
+    // Otherwise the ad is good to go.
+    return super.renderOutsideViewport();
+  }
+
+  /**
+   * To be overridden by network specific implementation indicating if element
+   * (and environment generally) are valid for sending XHR rqueries.
+   * @return {boolean} where element is valid and ad request should be sent.  If
+   *    false, no ad request is sent and slot will be collapsed if possible.
+   */
+  isValidElement() {
+    return true;
+  }
+
+  /**
+   * Prefetches and preconnects URLs related to the ad using adPreconnect
+   * registration which assumes ad request domain used for 3p is applicable.
+   * @override
+   */
+  preconnectCallback(unusedOnLayout) {
+    const preconnect = adPreconnect[this.element.getAttribute('type')];
+    // NOTE(keithwrightbos): using onLayout to indicate if preconnect should be
+    // given preferential treatment.  Currently this would be false when
+    // relevant (i.e. want to preconnect on or before onLayoutMeasure) which
+    // causes preconnect to delay for 1 sec (see custom-element#preconnect)
+    // therefore hard coding to true.
+    // NOTE(keithwrightbos): Does not take isValidElement into account so could
+    // preconnect unnecessarily, however it is assumed that isValidElement
+    // matches amp-ad loader predicate such that A4A impl does not load.
+    if (typeof preconnect == 'string') {
+      this.preconnect.url(preconnect, true);
+    } else if (preconnect) {
+      preconnect.forEach(p => {
+        this.preconnect.url(p, true);
+      });
+    }
+  }
+
+  /** @override */
+  onLayoutMeasure() {
+    if (this.intersectionObserver_) {
+      this.intersectionObserver_.fire();
+    }
+    if (this.layoutMeasureExecuted_ || !verifySignatureIsAvailable()) {
+      // onLayoutMeasure gets called multiple times.
+      return;
+    }
+    this.layoutMeasureExecuted_ = true;
+    if (isPositionFixed(this.element, this.getWin())) {
+      user.error('<amp-ad> is not allowed to be placed in elements with ' +
```

Comment:
Yeah, I'm changing this to an assert in layoutMeasure in [this PR](https://github.com/google/amphtml/pull/3) (which will land on this branch when Keith signs off).

The reason to do it in `onLayoutMeasure` instead of `layoutCallback` is to shortcut construction of the promise chain, generation of the XHR, etc.